docs: add documentation for all secure acl introduction work (#5640)
hashicorp/consul: website/source/api/acl/roles.html.md

---
layout: api
page_title: ACL Roles - HTTP API
sidebar_current: api-acl-roles
description: |-
  The /acl/role endpoints manage Consul's ACL Roles.
---

-> **1.5.0+:**  The APIs are available in Consul versions 1.5.0 and later. The documentation for the legacy ACL API is [here](/api/acl/legacy.html)

# ACL Role HTTP API

The `/acl/role` endpoints [create](#create-a-role), [read](#read-a-role),
[update](#update-a-role), [list](#list-roles) and [delete](#delete-a-role)  ACL roles in Consul.

> For more information about ACLs, please see the [ACL Guide](/docs/guides/acl.html).

Reviewer: This link will need to be updated, depending on where we document `roles`. If we decide to add it to the guides then we will need to link out to Learn. However, I thought we wanted to wait on adding roles to the guides since its a bit complicated for users who aren't using IdPs.

Author: That stub I left in place when cloning the Policy skeleton for now, but I agree that any links outside of `/api` are subject to change.

## Create a Role

This endpoint creates a new ACL role.

| Method | Path                         | Produces                   |
| ------ | ---------------------------- | -------------------------- |
| `PUT`  | `/acl/role`                  | `application/json`         |

The table below shows this endpoint's support for
[blocking queries](/api/index.html#blocking-queries),
[consistency modes](/api/index.html#consistency-modes),
[agent caching](/api/index.html#agent-caching), and

> [required ACLs](/api/index.html#acls).

Reviewer: *Warning* This the /api/index.html page might be updated with the filtering PR. It will affect the location of these sections (expect maybe ACLS)

| Blocking Queries | Consistency Modes | Agent Caching | ACL Required |
| ---------------- | ----------------- | ------------- | ------------ |
| `NO`             | `none`            | `none`        | `acl:write`  |

### Parameters

- `Name` `(string: <required>)` - Specifies a name for the ACL role. The name

>    can only contain alphanumeric characters as well as `-` and `_` and must be

Reviewer: ```suggestion
   can only contain alphanumeric characters, dashes `-` , underscores `_`,  and must be
```
If someone is using a screen reader it probably won't recognize just `_` or `-`

Author: That's actually text I cloned from the policies page (so it might be worth updating the same thing over there unrelated to my PR). I left a TODO because I need to update the validation text as it is more restrictive than policy names.

Reviewer: I think in this case its ok to be inconsistent for now, until we make bigger updates. 

We should maybe use an accessibility tool to catch all these things on the site and then make one big update.....

   unique. 
   
~> TODO(rb): update this.

- `Description` `(string: "")` - Free form human readable description of the role.

- `Policies` `(array<PolicyLink>)` - The list of policies that should
   be applied to the role. A PolicyLink is an object with an "ID" and/or "Name" field
   to specify a policy. With the PolicyLink, roles can be linked to policies either by the
   policy name or by the policy ID. When policies are linked by name they will be
   internally resolved to the policy ID. With linking roles internally by IDs,
   Consul enables policy renaming without breaking tokens.

> - `ServiceIdentities` `(array<ServiceIdentity>)` - The list of service

Reviewer: I think I still struggle a little with this concept. Maybe we can link out to the location where we define `service identity`

  identities that should be applied to the role. 

  - `ServiceName` `(string: <required>)` - The name of the service.

  - `Datacenters` `(array<string>)` - Specifies the datacenters the effective
    policy is valid within. When no datacenters are provided the 
    [effective policy](/docs/acl/acl-system.html#acl-service-identities) is
    valid in all datacenters including those which do not yet exist but may in
    the future.

### Sample Payload

```json
{
    "Name": "example-role",
    "Description": "Showcases all input parameters",
    "Policies": [
        {
            "ID": "783beef3-783f-f41f-7422-7087dc272765"
        },
        {
            "Name": "node-read"
        }
    ],
    "ServiceIdentities": [
        {
            "ServiceName": "web"
        },
        {
            "ServiceName": "db",
            "Datacenters": [
                "dc1"
            ]
        }
    ]
}
```

### Sample Request

> ```text

Reviewer: ```suggestion
```sh
```

                "dc1"
            ]
        }
    ],
    "Hash": "mBWMIeX9zyUTdDMq8vWB0iYod+mKBArJoAhj6oPz3BI=",
    "CreateIndex": 57,
    "ModifyIndex": 57
}
```

## Read a Role

This endpoint reads an ACL role with the given ID. If no role exists with the
given ID, a 404 is returned instead of a 200 response.

| Method | Path                         | Produces                   |
| ------ | ---------------------------- | -------------------------- |
| `GET`  | `/acl/role/:id`              | `application/json`         |

The table below shows this endpoint's support for
[blocking queries](/api/index.html#blocking-queries),
[consistency modes](/api/index.html#consistency-modes),
[agent caching](/api/index.html#agent-caching), and
[required ACLs](/api/index.html#acls).

| Blocking Queries | Consistency Modes | Agent Caching | ACL Required |
| ---------------- | ----------------- | ------------- | ------------ |
| `YES`            | `all`             | `none`        | `acl:read`   |

### Parameters

- `id` `(string: <required>)` - Specifies the UUID of the ACL role to
   read. This is required and is specified as part of the URL path.

### Sample Request

> ```text

Reviewer: ```suggestion
```sh
```

                "dc1"
            ]
        }
    ],
    "Hash": "mBWMIeX9zyUTdDMq8vWB0iYod+mKBArJoAhj6oPz3BI=",
    "CreateIndex": 57,
    "ModifyIndex": 57
}
```

## Read a Role by Name

This endpoint reads an ACL role with the given name. If no role exists with the
given name, a 404 is returned instead of a 200 response.

| Method | Path                         | Produces                   |
| ------ | ---------------------------- | -------------------------- |
| `GET`  | `/acl/role/name/:name`       | `application/json`         |

The table below shows this endpoint's support for
[blocking queries](/api/index.html#blocking-queries),
[consistency modes](/api/index.html#consistency-modes),
[agent caching](/api/index.html#agent-caching), and
[required ACLs](/api/index.html#acls).

| Blocking Queries | Consistency Modes | Agent Caching | ACL Required |
| ---------------- | ----------------- | ------------- | ------------ |
| `YES`            | `all`             | `none`        | `acl:read`   |

### Parameters

- `name` `(string: <required>)` - Specifies the Name of the ACL role to
   read. This is required and is specified as part of the URL path.

### Sample Request

> ```text

Reviewer: ```suggestion
```sh
```

   to specify a policy. With the PolicyLink, roles can be linked to policies either by the
   policy name or by the policy ID. When policies are linked by name they will be
   internally resolved to the policy ID. With linking roles internally by IDs,
   Consul enables policy renaming without breaking tokens.

- `ServiceIdentities` `(array<ServiceIdentity>)` - The list of service
  identities that should be applied to the role. 

  - `ServiceName` `(string: <required>)` - The name of the service.

  - `Datacenters` `(array<string>)` - Specifies the datacenters the effective
    policy is valid within. When no datacenters are provided the 
    [effective policy](/docs/acl/acl-system.html#acl-service-identities) is
    valid in all datacenters including those which do not yet exist but may in
    the future.

### Sample Payload

```json
{
    "ID": "8bec74a4-5ced-45ed-9c9d-bca6153490bb",
    "Name": "example-two",
    "Policies": [
        {
            "Name": "node-read"
        }
    ],
    "ServiceIdentities": [
        {
            "ServiceName": "db"
        }
    ]
}
```

### Sample Request

> ```text

Reviewer: ```suggestion
```sh
```

            "ServiceName": "db"
        }
    ],
    "Hash": "OtZUUKhInTLEqTPfNSSOYbRiSBKm3c4vI2p6MxZnGWc=",
    "CreateIndex": 14,
    "ModifyIndex": 28
}
```

## Delete a Role

This endpoint deletes an ACL role.

| Method   | Path                      | Produces                   |
| -------- | ------------------------- | -------------------------- |
| `DELETE` | `/acl/role/:id`           | `application/json`         |

Even though the return type is application/json, the value is either true or
false indicating whether the delete succeeded.

The table below shows this endpoint's support for
[blocking queries](/api/index.html#blocking-queries),
[consistency modes](/api/index.html#consistency-modes),
[agent caching](/api/index.html#agent-caching), and
[required ACLs](/api/index.html#acls).

| Blocking Queries | Consistency Modes | Agent Caching | ACL Required |
| ---------------- | ----------------- | ------------- | ------------ |
| `NO`             | `none`            | `none`        | `acl:write`  |

### Parameters

- `id` `(string: <required>)` - Specifies the UUID of the ACL role to
  delete. This is required and is specified as part of the URL path.

### Sample Request

> ```text

Reviewer: ```suggestion
```sh
```

$ curl -X DELETE \
    http://127.0.0.1:8500/v1/acl/role/8f246b77-f3e1-ff88-5b48-8ec93abf3e05
```

### Sample Response
```json
true
```

## List Roles

This endpoint lists all the ACL roles.

| Method | Path                         | Produces                   |
| ------ | ---------------------------- | -------------------------- |
| `GET`  | `/acl/roles`                 | `application/json`         |

The table below shows this endpoint's support for
[blocking queries](/api/index.html#blocking-queries),
[consistency modes](/api/index.html#consistency-modes),
[agent caching](/api/index.html#agent-caching), and
[required ACLs](/api/index.html#acls).

| Blocking Queries | Consistency Modes | Agent Caching | ACL Required |
| ---------------- | ----------------- | ------------- | ------------ |
| `YES`            | `all`             | `none`        | `acl:read`   |

## Parameters

- `policy` `(string: "")` - Filters the role list to those roles that are
  linked with the specific policy ID.

## Sample Request

> ```text

Reviewer: ```suggestion
```sh
```
